jlm/mlir/frontend/MlirToRvsdg.cpp

```diff
@@ -0,0 +1,256 @@
+  std::unique_ptr<mlir::Block> block = std::make_unique<mlir::Block>();
+  auto result = mlir::parseSourceFile(filePath.to_str(), block.get(), config);
+  if (result.failed())
+  {
+    throw util::error("Parsing MLIR input file failed.");
+  }
+  return block;
+}
+
+std::unique_ptr<llvm::RvsdgModule>
+MlirToRvsdg::convertMlir(std::unique_ptr<mlir::Block> & block)
+{
+  std::string dataLayout;
+  std::string targetTriple;
+  util::filepath sourceFileName("");
+  auto rvsdgModule = llvm::RvsdgModule::Create(sourceFileName, targetTriple, dataLayout);
+  auto & graph = rvsdgModule->Rvsdg();
+  auto root = graph.root();
+
+  convertBlock(*block.get(), *root);
+
+  return rvsdgModule;
+}
+
+std::unique_ptr<std::vector<jlm::rvsdg::output *>>
+MlirToRvsdg::convertRegion(mlir::Region & region, rvsdg::region & rvsdgRegion)
+{
+  // MLIR use blocks as the innermost "container"
+  // In the RVSDG Dialect a region should contain one and only one block
+  JLM_ASSERT(region.getBlocks().size() == 1);
+  return convertBlock(region.front(), rvsdgRegion);
+}
+
+std::unique_ptr<std::vector<jlm::rvsdg::output *>>
+MlirToRvsdg::convertBlock(mlir::Block & block, rvsdg::region & rvsdgRegion)
+{
+  mlir::sortTopologically(&block);
+
+  // Create an RVSDG node for each MLIR operation and store each pair in a
+  // hash map for easy lookup of corresponding RVSDG nodes
+  std::unordered_map<mlir::Operation *, rvsdg::node *> operations;
+  for (mlir::Operation & mlirOp : block.getOperations())
+  {
+    std::vector<const jlm::rvsdg::output *> inputs;
+    for (mlir::Value operand : mlirOp.getOperands())
+    {
+      if (mlir::Operation * producer = operand.getDefiningOp())
+      {
+        size_t index = getOperandIndex(producer, operand);
+        inputs.push_back(operations[producer]->output(index));
+      }
+      else
+      {
+        // If there is no defining op, the Value is necessarily a Block argument.
+        auto blockArg = operand.cast<mlir::BlockArgument>();
+        inputs.push_back(rvsdgRegion.argument(blockArg.getArgNumber()));
+      }
+    }
+
+    if (rvsdg::node * node = convertOperation(mlirOp, rvsdgRegion, inputs))
+    {
+      operations[&mlirOp] = node;
+    }
+  }
+
+  // The results of the region/block are encoded in the terminator operation
+  auto terminator = block.getTerminator();
+  std::unique_ptr<std::vector<jlm::rvsdg::output *>> results =
+      std::make_unique<std::vector<jlm::rvsdg::output *>>();
+  for (mlir::Value operand : terminator->getOperands())
+  {
+    if (mlir::Operation * producer = operand.getDefiningOp())
+    {
+      size_t index = getOperandIndex(producer, operand);
+      results->push_back(operations[producer]->output(index));
+    }
+    else
+    {
+      // If there is no defining op, the Value is necessarily a Block argument.
+      auto blockArg = operand.cast<mlir::BlockArgument>();
+      results->push_back(rvsdgRegion.argument(blockArg.getArgNumber()));
```

Comment:
The same here. Could we insert an JLM_ASSERT to check that it is actually a block argument?

Reply:
Same as earlier.

Reply:
see above.

Reply:
Added